NitroxClient/GameLogic/Vehicles.cs

```diff
@@ -106,5 +120,170 @@ private void OnVehiclePrefabLoaded(GameObject prefab, string guid, Vector3 spawn
+
+                    if (player.IsPresent()) // Get Remote Player Inside == vehicle.pilotId  Remove From Vehicle Before Destroy
+                    {
+                        RemotePlayer playerInstance = player.Get();
+                        playerInstance.SetVehicle(null);
+                        playerInstance.SetSubRoot(null);
+                        playerInstance.SetPilotingChair(null);
+                        playerInstance.AnimationController.UpdatePlayerAnimations = true;
+                    }
+                }
+
+                if (vehicle.GetPilotingMode()) //Check Local Object Have Player inside
+                {
+                    vehicle.ReflectionCall("OnPilotModeEnd", false, false, null);
+                    if (!Player.main.ToNormalMode(true))
+                    {
+                        Player.main.ToNormalMode(false);
+                        Player.main.transform.parent = null;
+                    }
+                }
+
+                if (vehicle.gameObject != null) // Destroy Vehicle
+                {
+                    if (vehicle.destructionEffect)
+                    {
+                        GameObject gameObject = UnityEngine.Object.Instantiate<GameObject>(vehicle.destructionEffect);
+                        gameObject.transform.position = vehicle.transform.position;
+                        gameObject.transform.rotation = vehicle.transform.rotation;
+                    }
+                    UnityEngine.Object.Destroy(vehicle.gameObject);
+                }
+            }
+        }
+
+        public void Add(Vehicle vehicle) //Add Vehicle Fron Network Packet
+        {
+            string guid;
+            Vector3 position;
+            Quaternion rotation;
+            Vector3 velocity;
+            Vector3 angularVelocity;
+            TechType techType;
+            float steeringWheelYaw = 0f, steeringWheelPitch = 0f;
+            bool appliedThrottle = false;
+
+            guid = GuidHelper.GetGuid(vehicle.gameObject);
+            position = vehicle.gameObject.transform.position;
+            rotation = vehicle.gameObject.transform.rotation;
+            techType = CraftData.GetTechType(vehicle.gameObject);
+
+            Rigidbody rigidbody = vehicle.gameObject.GetComponent<Rigidbody>();
+
+            velocity = rigidbody.velocity;
+            angularVelocity = rigidbody.angularVelocity;
+
+            // Required because vehicle is either a SeaMoth or an Exosuit, both types which can't see the fields either.
+            steeringWheelYaw = (float)vehicle.ReflectionGet<Vehicle, Vehicle>("steeringWheelYaw");
+            steeringWheelPitch = (float)vehicle.ReflectionGet<Vehicle, Vehicle>("steeringWheelPitch");
+
+            // Vehicles (or the SeaMoth at least) do not have special throttle animations. Instead, these animations are always playing because the player can't even see them (unlike the cyclops which has cameras).
```

Comment:
It looks like this code was copied from PlayerMovement.  Can we move that function here and have player movement call into vehicle instead? (we might have to do some work to make it generic) That way we don't have redundant code.

Reply:
And then, most if not all of the movement/animation code is of no importance when creating the vehicle. This can all be kept in the movement update codepath.